### DISCUSSION TITLE
Is grab() thread-safe?

### DISCUSSION THREAD

**Discussion by Zjaun** [bytedeco/javacv#2118] [Q&A] [answered]
I have been developing a project for a while now (color conversions and dithering), and I wanted to assess the performance of each project by creating a separate thread. One thread for the raw input, one thread for a filter, and one thread for dithering. This is how I implement that: (side note, the `CustomVideoInputFrameGrabber` implementation is the one I created on [this](https://github.com/bytedeco/javacv/discussions/2103#discussioncomment-7174107) post)

```
public static void main(String[] args) throws Exception {
          FrameGrabber grabber = new CustomVideoInputFrameGrabber(1);
          grabber.setImageWidth(640);
          grabber.setImageHeight(360);
          grabber.start();
          Demo webcam = new Demo("Webcam", 0, grabber);
          Demo directRGB = new Demo("Direct Palette Conversion (1 Thread)", 1, grabber);
          Demo directDither = new Demo("Dithered-FSD (1 Thread)", 2, grabber);
          Thread webcamThread = new Thread(webcam);
          Thread directRGBThread = new Thread(directRGB);
          Thread directDitherThread = new Thread(directDither);
          webcamThread.start();
          directRGBThread.start();
          directDitherThread.start();
}
```

Demo class implementation:
```
public class Demo implements Runnable {

    private String windowName;
    private int mode;
    private FrameGrabber grabber;

    public Demo(String windowName, int mode, FrameGrabber grabber) {
        this.windowName = windowName;
        this.mode = mode;
        this.grabber = grabber;
    }

    @Override
    public void run() {
        Java2DFrameConverter converter = new Java2DFrameConverter();
        CanvasFrame window = new CanvasFrame(windowName);
        switch (mode) {
            case 0:
                while (true) {
                    try {
                        window.showImage(grabber.grab());
                    } catch (Exception e) {

                    }
                }
            case 1:
                while (true) {
                    try {
                        window.showImage(ColorQuantizer.toRGB(converter.convert(grabber.grab())));
                    } catch (Exception e) {

                    }
                }
            case 2:
                while (true) {
                    try {
                        window.showImage(ColorQuantizer.dither(converter.convert(grabber.grab())));
                    } catch (Exception e) {

                    }
                }
            default:
                System.out.println("Mode is outside the range! " + mode);
                return;
        }
    }

}
```
Where the second parameter of constructor of the `Demo` class represents what the `CanvasFrame` would represent (Raw, Filtered, or Dithered).

I sometimes have some success running all of these three simultaneously (see video)

https://github.com/bytedeco/javacv/assets/91415509/c17475c2-4094-4247-b4d2-e1bfbeeaf8ea

However, I sometimes stumble upon an access violation exception, stating that the .dll file `jnivideoInputLib.dll` is at fault.

[hs_err_pid10728.log](https://github.com/bytedeco/javacv/files/13227310/hs_err_pid10728.log)

I can mitigate the issue by creating another synchronized method that contains the grabber.grab() function, but I still wonder why on some occassions it works, but sometimes it doesn't.

**Comment by saudet** [comment, marked as answer]
Some video drivers don't like to be called from different threads. It's not really a synchronization issue, just don't use the grabber from multiple threads.
